cpp/harvest/marc_issn_lookup.cc

```diff
@@ -0,0 +1,171 @@
+    bool is_found;
+
+    SubFieldInfo(bool b) { this->is_found = b; }
+
+    void Builder(MARC::Record &record) {
+        for (auto &field : record) {
+            if (field.getTag() == "001")
+                this->w = "(DE-627)" + field.getContents();
+
+            if (field.getTag() == "773")
+                this->x = field.getContents();
+
+            if (field.getTag() == "245") {
+                MARC::Subfields subfields(field.getSubfields());
+                std::string subfield_a(subfields.getFirstSubfieldWithCode('a'));
+                std::string subfield_b(subfields.getFirstSubfieldWithCode('b'));
+                if (subfield_a != "" && subfield_b != "")
+                    this->t = subfield_a + " " + subfield_b;
+                else if (subfield_a != "" && subfield_b == "")
+                    this->t = subfield_a;
+                else if (subfield_a == "" && subfield_b != "")
+                    this->t = subfield_b;
+                else
+                    this->t = "";
+            }
+        }
+        this->i = "In:";
+        this->is_found = true;
+    }
+};
+
+bool IsOnlineISSN(const MARC::Record &record) {
+    for (auto &field : record) {
+        if (field.getTag() == "300") {
+            const std::string sub_field_a(field.getFirstSubfieldWithCode('a'));
+            if (sub_field_a == "Online-Ressource")
+                return true;
+            else
+                return false;
+        }
+    }
+    return false;
+}
+
+SubFieldInfo LookupISSNInDataSouce(const std::string input_journal_filename, const std::string issn) {
+    auto data_source(MARC::Reader::Factory(input_journal_filename));
+    SubFieldInfo sub_field_info = SubFieldInfo(false);
+
+    while (MARC::Record record = data_source->read()) {
+        for (auto &field : record) {
+            if (field.getTag() == "022") {
+                const std::string sub_field_a(field.getFirstSubfieldWithCode('a'));
+                if ((sub_field_a != "") && (sub_field_a == issn)) {
+                    sub_field_info.Builder(record);
+                    // the issn is online version, no need to check further
+                    if (IsOnlineISSN(record))
+                        return sub_field_info;
+                } else {
+                    // maybe the issn is in subfield l
+                    const std::string sub_field_l(field.getFirstSubfieldWithCode('l'));
+                    if ((sub_field_l != "") && (sub_field_l == issn)) {
+                        sub_field_info.Builder(record);
+                        // the issn is online version, no need to check further
+                        if (IsOnlineISSN(record))
+                            return sub_field_info;
+                    }
+                }
+            }
+        }
+    }
+    return sub_field_info;
+}
+
+void UpdateSubfield(MARC::Subfields &subfields, const SubFieldInfo &sub_field_info) {
+    if (!subfields.replaceFirstSubfield('i', sub_field_info.i))
```

Comment:
I think there is an existing function `insertOrReplaceSubfield` on a MARC::Field which should be more efficient here.